cmd/snapshot-controller/main.go

```diff
@@ -59,6 +63,45 @@ var (
 	version = "unknown"
 )
 
+// Checks that the VolumeSnapshot v1 CRDs exist.
+func ensureCustomResourceDefinitionsExist(kubeClient *kubernetes.Clientset, client *clientset.Clientset) error {
+	condition := func() (bool, error) {
+		var err error
+		_, err = kubeClient.CoreV1().Namespaces().Get(context.TODO(), "kube-system", metav1.GetOptions{})
+		if err != nil {
+			// only execute list VolumeSnapshots if the kube-system namespace exists
+			_, err = client.SnapshotV1().VolumeSnapshots("kube-system").List(context.TODO(), metav1.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed to list v1 volumesnapshots with error=%+v", err)
+				return false, nil
+			}
+		}
+		_, err = client.SnapshotV1().VolumeSnapshotClasses().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			klog.Errorf("Failed to list v1 volumesnapshotclasses with error=%+v", err)
+			return false, nil
+		}
+		_, err = client.SnapshotV1().VolumeSnapshotContents().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			klog.Errorf("Failed to list v1 volumesnapshotcontents with error=%+v", err)
+			return false, nil
+		}
+		return true, nil
+	}
+
+	// with a Factor of 1.5 we wait up to 7.5 seconds (the 10th attempt)
+	backoff := wait.Backoff{
+		Duration: 100 * time.Millisecond,
+		Factor:   1.5,
+		Steps:    10,
+	}
+	if err := wait.ExponentialBackoff(backoff, condition); err != nil {
```

Comment:
I don't understand, why need to wait?

Reply:
Wait until the CRDs are installed.

Reply:
if crash immediately and not retry, when the container restart again, it will check agian?

Reply:
yes however we'd like to have a faster startup time, if the container exits quickly and goes into `CrashLoopBackOff` it'll be restarted every 10s, 20s, 40s and so on, during the first run instead of failing quickly it might be possible that the CRDs are ready after a few seconds and in one of these 10 runs the wait condition might succeed

Reply:
@mauriciopoppe  copy that, thanks.